src/ias_proxy/src/ias_proxy_server.rs

```diff
@@ -0,0 +1,444 @@
+/*
+ Copyright 2018 Intel Corporation
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+------------------------------------------------------------------------------
+*/
+
+extern crate futures;
+extern crate hyper;
+extern crate ias_client;
+extern crate serde;
+extern crate serde_json;
+
+use ias_proxy_config::IasProxyConfig;
+use lru_cache::LruCache;
+use self::futures::{Future,
+                    future};
+use self::hyper::{Body,
+                  Error,
+                  header::{HeaderMap,
+                           HeaderValue},
+                  Method,
+                  Request,
+                  Response,
+                  Server,
+                  service::service_fn,
+                  StatusCode};
+use self::ias_client::{client_utils,
+                       client_utils::{ClientError,
+                                      ClientResponse},
+                       ias_client::IasClient};
+use std::{borrow::Borrow,
+          net::SocketAddr,
+          str::FromStr,
+          sync::Mutex};
+use utils::read_file_as_string;
+
+/// type definition for response sent from web server
+type ResponseBox = Box<Future<Item=Response<Body>, Error=Error> + Send>;
+
+/// structure defining IAS proxy server
+pub struct IasProxyServer {
+    ias_proxy_ip: String,
+    ias_proxy_port: String,
+    ias_client: Box<IasClient>,
+}
+
+/// Request body from client, proxy server may deserialize the reuqest in order to get quote if
+/// request is for attestation verification report.
+#[derive(Deserialize)]
+struct IasAVRRequestBody {
+    #[serde(rename = "isvEnclaveQuote")]
+    isv_enclave_quote: String,
+    #[serde(rename = "pseManifest")]
+    pse_manifest: String,
+    nonce: u64,
+}
+
+/// ClientResponse decoded information stored in cache
+#[derive(Debug, Clone)]
+struct IasResponse {
+    body_string: String,
+    header_map: HeaderMap,
+}
+
+lazy_static! {
+    static ref sig_rl_cache: Mutex<LruCache<String, IasResponse>> =
+        Mutex::new(LruCache::new(None));
```

Comment:
Storing of reference is not possible, LruCache internally is a hashmap implementation which takes ownership of key and value. But I can think of returning reference of value stored instead of cloning it before send.